cpp/comms/src/cuML_std_comms_impl.cpp

```diff
@@ -0,0 +1,417 @@
+    case MLCommon::cumlCommunicator::DOUBLE:
+      return sizeof(double);
+  }
+}
+
+ncclDataType_t getNCCLDatatype(
+  const cumlStdCommunicator_impl::datatype_t datatype) {
+  switch (datatype) {
+    case MLCommon::cumlCommunicator::CHAR:
+      return ncclChar;
+    case MLCommon::cumlCommunicator::UINT8:
+      return ncclUint8;
+    case MLCommon::cumlCommunicator::INT:
+      return ncclInt;
+    case MLCommon::cumlCommunicator::UINT:
+      return ncclUint32;
+    case MLCommon::cumlCommunicator::INT64:
+      return ncclInt64;
+    case MLCommon::cumlCommunicator::UINT64:
+      return ncclUint64;
+    case MLCommon::cumlCommunicator::FLOAT:
+      return ncclFloat;
+    case MLCommon::cumlCommunicator::DOUBLE:
+      return ncclDouble;
+  }
+}
+
+ncclRedOp_t getNCCLOp(const cumlStdCommunicator_impl::op_t op) {
+  switch (op) {
+    case MLCommon::cumlCommunicator::SUM:
+      return ncclSum;
+    case MLCommon::cumlCommunicator::PROD:
+      return ncclProd;
+    case MLCommon::cumlCommunicator::MIN:
+      return ncclMin;
+    case MLCommon::cumlCommunicator::MAX:
+      return ncclMax;
+  }
+}
+}  // namespace
+
+
+bool ucx_enabled(){
+  return UCX_ENABLED;
+}
+
+/**
+ * @brief Underlying comms, like NCCL and UCX, should be initialized and ready for use,
+ * and maintained, outside of the cuML Comms lifecycle. This allows us to decouple the
+ * ownership of the actual comms from cuml so that they can also be used outside of cuml.
+ *
+ * For instance, nccl-py can be used to bootstrap a ncclComm_t before it is
+ * used to construct a cuml comms instance. UCX endpoints can be bootstrapped
+ * in Python using ucx-py, before being used to construct a cuML comms instance.
+ */
+#ifdef WITH_UCX
+void inject_comms(cumlHandle &handle, ncclComm_t comm, ucp_worker_h ucp_worker,
+                  std::shared_ptr<ucp_ep_h *> eps, int size, int rank) {
+  auto communicator = std::make_shared<MLCommon::cumlCommunicator>(
+    std::unique_ptr<MLCommon::cumlCommunicator_iface>(
+      new cumlStdCommunicator_impl(comm, ucp_worker, eps, size, rank)));
+  handle.getImpl().setCommunicator(communicator);
+}
+#endif
+
+void inject_comms(cumlHandle &handle, ncclComm_t comm, int size, int rank) {
+  auto communicator = std::make_shared<MLCommon::cumlCommunicator>(
+    std::unique_ptr<MLCommon::cumlCommunicator_iface>(
+      new cumlStdCommunicator_impl(comm, size, rank)));
+  handle.getImpl().setCommunicator(communicator);
+}
+
+void inject_comms_py_coll(cumlHandle *handle, ncclComm_t comm, int size,
+                          int rank) {
+  inject_comms(*handle, comm, size, rank);
+}
+
+void inject_comms_py(ML::cumlHandle *handle, ncclComm_t comm, void *ucp_worker,
+                     void *eps, int size, int rank) {
+
+
+#ifdef WITH_UCX
+  ucp_worker_print_info((ucp_worker_h)ucp_worker, stdout);
+
+  // Make shared_ptr for endpoints
+  std::shared_ptr<ucp_ep_h *> eps_sp =
+    std::make_shared<ucp_ep_h *>(new ucp_ep_h[size]);
```

Comment:
In case this shared_ptr is really needed: I might miss something but without a custom delete function I think `std::make_shared` will call `delete` and not ` delete []`.

Reply:
Discussed offline, but `ucp_ep_h*` is actually an array of pointers and not a `ucp_ep_h*` array.

Reply:
I've done some research and tried some alternate design approaches. Unfortunately, Cython makes this nontrivial. I'm using a little trick in Cython to be able to pass pointers around in the native Python layer by casting them to `size_t` (a type natively convertable between Cython & Python which is large enough to store a pointer). I never actually get an instance of a ucp endpoint object in Cython and I'm only passing their pointers down to C++. Because of this, I don't have an array of `ucp_ep_h` objects, I have an array of `size_t` pointers which each need to be casted back to `ucp_ep_h` pointers, hence the coupling to Python being done with `inject_comms_py` and not `inject_comms`, which is part of the user-facing public API. This is also why I separated the header files. This is very similar to how Akshay build `ucx-py`, actually.

Looking into the `shared_ptr` documentation, the `inject_comms_py` function should be fine using a `shared_ptr` here it because I'm constructing a new array with the pointers to objects that are owned by another object. The only object that needs to be released is the array itself, not the elements within it. The other option would be passing in the `size_t*` from Python directly to the `inject_comms` function but that would cause Python-specific details to leak into the user-facing comms API, which would not be good. 

We could argue this is already happening with the use of the shared pointer, but I don't see a reasonable way around it at the present time.

Reply:
Thanks. In this case I agree it makes sense to keep the design as it it. It would be nice if Cython would offer a nicer way to do this though :-(